# Exploring Hacker News Posts

## In this Project we will work on [dataset](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts) \ Which is from [Hacker News](https://news.ycombinator.com/) that has details about posts on its forum.

## This Dataset has various attributes(columns):

`id`: the unique identifier from Hacker News for the post

`title`: the title of the post

`url`: the URL that the posts links to, if the post has a URL

`num_points`: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes

`num_comments`: the number of comments on the post

`author`: the username of the person who submitted the post

`created_at`: the date and time of the post's submission

## The Goal here is to determine 2 types of post:

* Do `Ask Hn` or `Show HN` recieve more comments on average?
* Do posts created at a certain time recieve more comments on average?

In [3]:
# Importing necessary modules
import csv
# for calculating comments by hour we need datetime module 
import datetime as dt

In [4]:
openfile = open("hacker_news.csv")

hn = list(csv.reader(openfile))
# Seperating header of dataset from `hn` list of list
headers = hn[:1]
# Assigning rest of dataset back to hn
hn = hn[1:]
#Dislaying both of them
print(headers)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting `Ask Hn` and `Show Hn` Posts

In [5]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
        
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
print(ask_posts[:5])
print(show_posts[:5])
print(other_posts[:5])   
print(len(hn))

1744
1162
17194
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.i

## Finding total number comments in both `Ask Hn` and `Show Hn` posts

Total number comments in `Ask Hn`:

In [6]:
total_ask_comments = 0
count = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    count += 1
avg_ask_comments = total_ask_comments/count 
print(total_ask_comments)
print(round(avg_ask_comments,3))

24483
14.038


Total number comments in `Show Hn`:

In [7]:
total_show_comments = 0
count = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    count += 1
avg_show_comments = total_show_comments/count 
print(total_show_comments)
print(round(avg_show_comments,3))

11988
10.317


* The total number of ask comments `total_ask_comments` : 24483
* The Average number of ask comments `avg_ask_comments` : 14.038
* The total number of show comments `total_show_comments` : 11988
* The Average number of show comments `avg_show_comments` : 10.317

#### Hence Average number of `avg_ask_comments` are greater than `avg_show_comments`

## Finding the Number of Ask Posts and Comments by Hour Created

##### Here we will determine if ask posts created at a certain time are more likely to attract comments, for that we will do following steps:

1. Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

### Step 1:

* For calculating the number of ask posts and comments by hour we will use `datetime` module
* From datetime module we will use `datetime.strptime()` method to parse dates stored as strings and return datetime objects

In [8]:
#creating empty list
result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at,num_comments])

    
print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [9]:
#creating 2 dictionaries 
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comment = row[1]
    date_t = dt.datetime.strptime(date,"%m/%d/%Y  %H:%M")
    hour = date_t.strftime("%H")
    #print(time)
    counts_by_hour[hour] = counts_by_hour.get(hour,0)+1
    #print(comment)
    comments_by_hour[hour] = comments_by_hour.get(hour,comment)+comment
#print(comments_by_hour)
print(counts_by_hour)

comments_by_hour

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


{'09': 257,
 '13': 1282,
 '10': 794,
 '14': 1419,
 '16': 1831,
 '23': 544,
 '12': 691,
 '17': 1147,
 '15': 4478,
 '21': 1749,
 '20': 1724,
 '02': 1384,
 '18': 1441,
 '03': 422,
 '05': 493,
 '19': 1191,
 '01': 716,
 '22': 481,
 '08': 497,
 '04': 340,
 '00': 457,
 '06': 398,
 '07': 269,
 '11': 643}

## Calculating the Average Number of Comments for `Ask Hn` Posts by Hour

* `counts_by_hour` : contains the number of ask posts created during each hour of the day.
* `comments_by_hour` : contains the corresponding number of comments ask posts created at each hour received.

#### Now we will use these two dictionaries to calculate the average number of comments for posts created during each hour of the day

In [10]:
avg_by_hour = []

for k in comments_by_hour:
    avg = round(comments_by_hour[k]/counts_by_hour[k],3)
    avg_by_hour.append([k,avg])

print(avg_by_hour)

[['09', 5.711], ['13', 15.082], ['10', 13.458], ['14', 13.262], ['16', 16.954], ['23', 8.0], ['12', 9.466], ['17', 11.47], ['15', 38.603], ['21', 16.046], ['20', 21.55], ['02', 23.862], ['18', 13.22], ['03', 7.815], ['05', 10.717], ['19', 10.827], ['01', 11.933], ['22', 6.775], ['08', 10.354], ['04', 7.234], ['00', 8.309], ['06', 9.045], ['07', 7.912], ['11', 11.086]]


## Sorting and Printing Values from a List of Lists : `avg_by_hour`

In [11]:
#creating list of list avg_by_hour with swapped colums
swap_avg_by_hour = []

for h,c in avg_by_hour:
    swap_avg_by_hour.append([c,h])
print(swap_avg_by_hour)

[[5.711, '09'], [15.082, '13'], [13.458, '10'], [13.262, '14'], [16.954, '16'], [8.0, '23'], [9.466, '12'], [11.47, '17'], [38.603, '15'], [16.046, '21'], [21.55, '20'], [23.862, '02'], [13.22, '18'], [7.815, '03'], [10.717, '05'], [10.827, '19'], [11.933, '01'], [6.775, '22'], [10.354, '08'], [7.234, '04'], [8.309, '00'], [9.045, '06'], [7.912, '07'], [11.086, '11']]


In [12]:
#sorting swapswap_avg_by_hour in desecending order

sorted_swap = sorted(swap_avg_by_hour,key = lambda x :x[0],reverse = True )
print(sorted_swap)

[[38.603, '15'], [23.862, '02'], [21.55, '20'], [16.954, '16'], [16.046, '21'], [15.082, '13'], [13.458, '10'], [13.262, '14'], [13.22, '18'], [11.933, '01'], [11.47, '17'], [11.086, '11'], [10.827, '19'], [10.717, '05'], [10.354, '08'], [9.466, '12'], [9.045, '06'], [8.309, '00'], [8.0, '23'], [7.912, '07'], [7.815, '03'], [7.234, '04'], [6.775, '22'], [5.711, '09']]


### Printing "Top 5 Hours for Ask Posts Comments"

In [13]:
for c,h in sorted_swap[:5]:
    
    h = dt.datetime.strptime(h,"%H")
    
    #h = dt.time('%)
    #print(type(h))
    only_h = h.time()
    #print(type(only_h))
    only_h = only_h.strftime("%H:%M")
    #print(type(only_h))
    print("{} {:.2f} average comments per post".format(only_h,c))

15:00 38.60 average comments per post
02:00 23.86 average comments per post
20:00 21.55 average comments per post
16:00 16.95 average comments per post
21:00 16.05 average comments per post


## Conclusions

### These are the Top 5 hours in which there are higher chances to get comments on your Ask Posts:


* 15:00 38.60 average comments per post
* 02:00 23.86 average comments per post
* 20:00 21.55 average comments per post
* 16:00 16.95 average comments per post
* 21:00 16.05 average comments per post

#### As we know time given in our [**dataset] is mentioned as Eastern Daylight Time (EDT) you can refer to [Documentation](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts)

* As we know The Eastern Daylight Time (EDT) is 9.5 hours behind of India Standard Time (IST).
* To comment on `Ask Hn` Post in India the best Hour would be 12.30 in The night or 00:30 in the Night

So, to maximize the potential number of comments it makes sence to create an ask post and do it at Midnight past 30 mins